### PR TITLE
Smallest update in `dependency-services report`

### PR DESCRIPTION
**Pull request by sigurdm**
The pub-side of solving: https://github.com/dependabot/dependabot-core/issues/5391

`dependency-services report` is fed a list of packages package (name+hosted url) and for each a list of disallowed (insecure) version-ranges.

It will then run the solver on a modified pubspec.yaml with all versions bounded to at least the current version. The disallowed ranges will be fed into the solver as additional constraints. The solver will be run in 'downgrade' mode, ensuring that the update is as small as possible.

Some open questions:
* Should we allow for also downgrading the target package to a not-affected version?
 - That would require another strategy than "constrain and downgrade"...
* Is it enough to handle only hosted packages?
  - I don't think we will have security advisories for git packages...
 
There is currently no error-handling (if we cannot resolve, we simply report no smallestUpdate) this might be ok for a first iteration.

### CODE REVIEW

File: lib/src/command/dependency_services.dart [hunk 1]
```diff
@@ -200,6 +98,31 @@ class DependencyServicesReportCommand extends PubCommand {
         singleBreakingVersion = singleBreakingPackagesResult
             ?.firstWhereOrNull((element) => element.name == package.name);
       }
+      PackageId? smallestUpgrade;
+      final disallowedRanges = disallowed
+          .firstWhereOrNull((disallowed) => disallowed.ref == package.toRef());
+      if (disallowedRanges != null &&
+          !disallowedRanges.constraints.every(
+            (element) => !element.constraint.allows(package.version),
+          )) {
```
Comment (jonasfj) [inline lib/src/command/dependency_services.dart:107]: Instead of not every not, could we do: `any`?

**Review by jonasfj — APPROVED** [review]

**Comment by jonasfj** [comment]
> Should we allow for also downgrading the target package to a not-affected version?

I propose we don't

**Comment by jonasfj** [comment]
> Is it enough to handle only hosted packages?

Yes, for now I think that's good enough.

It's possible that dependabot will have advisory information for git-dependencies, though, I'm not sure how we would handle that.

**Comment by jonasfj** [comment]
> There is currently no error-handling (if we cannot resolve, we simply report no smallestUpdate) this might be ok for a first iteration.

What is the alternative? I'm not sure I understand what is missing :D

**Comment by jonasfj** [comment]
Should `disallowed` affect other resolutions? And should we support `disallowed` containing multiple packages?

**Comment by sigurdm** [comment]
> What is the alternative? I'm not sure I understand what is missing :D

I was thinking of some sort of communicating the solve-trace back to dependabot, so the user would know why the security-update failed.

I don't know that dependabot has facilities for this.

**Comment by sigurdm** [comment]
> Should disallowed affect other resolutions? 

Good question - not really sure what we want to do here. It will most likely not be used by our dependabot integration, but seems to be the right thing to do in general.

> And should we support disallowed containing multiple packages?
We do support this already. Should probably write a test :).

**Comment by sigurdm** [comment]
> > Should disallowed affect other resolutions?

Actually thinking about this again, this could probably fill in our missing feature of handling "ignored" versions correctly.

**Comment by sigurdm** [comment]
I changed the solver to take a list of constraints instead of incompatibilities to not expose solver internals.

I made all the resolutions be affected by the forbidden dependencies, such that we can also use it for ignored versions.

Also now you can pass a "target" package name to `report` and it will only compute the report for that package. (This is to save resources when we cannot cache the report between invocations).

**Comment by sigurdm** [comment]
PTAL

File: lib/src/command/dependency_services.dart (hunk 1 above)
Comment (sigurdm) [reply lib/src/command/dependency_services.dart]: Done (ish)

File: lib/src/command/dependency_services.dart [hunk 2]
```diff
@@ -670,3 +580,228 @@ bool _lockFileHasContentHashes(dynamic lockfile) {
...
+  if (disallowedList is! List<Object?>) {
+    throw FormatException('Disallowed should be a list of maps');
+  }
+  final result = <_DisallowedPackageRanges>[];
+  for (final disallowed in disallowedList) {
+    if (disallowed is! Map) {
+      throw FormatException('Disallowed should be a list of maps');
+    }
+    final ref = PackageRef(
+      disallowed['name'] as String,
```
[the remark this answers is not in the document]
Comment (sigurdm) [reply lib/src/command/dependency_services.dart]: Done

File: lib/src/solver.dart [hunk 3]
```diff
@@ -27,21 +29,29 @@ export 'solver/type.dart';
...
+///
+/// [extraConstraints] can contain a list of extra constraints for this
+/// resolution.
 Future<SolveResult> resolveVersions(
   SolveType type,
   SystemCache cache,
   Package root, {
   LockFile? lockFile,
   Iterable<String> unlock = const [],
+  Iterable<ConstraintAndCause>? extraConstraints,
```
[the remark this answers is not in the document]
Comment (sigurdm) [reply lib/src/solver.dart:42]: Done

File: lib/src/solver/version_solver.dart [hunk 4]
```diff
@@ -568,3 +580,13 @@ class VersionSolver {
     log.solver(prefixLines(message, prefix: '  ' * _solution.decisionLevel));
   }
 }
+
+// An additional constraint to a version resolution.
+class ConstraintAndCause {
+  /// Stated in the **negative** like constraints in the pubspec, meaning that
+  /// to forbid a version you must do
+  /// `VersionConstraint.any.difference(version)`.
```
[the remark this answers is not in the document]
Comment (sigurdm) [reply lib/src/solver/version_solver.dart:601]: Yeah, that was not well explained. Tried to clariy

File: lib/src/solver/version_solver.dart [hunk 5]
```diff
@@ -568,3 +580,13 @@ class VersionSolver {
...
+  final PackageRange range;
+  final String? cause;
+  ConstraintAndCause(this.range, this.cause);
```
[the remark this answers is not in the document]
Comment (sigurdm) [reply lib/src/solver/version_solver.dart:604]: Done

File: lib/src/command/dependency_services.dart [hunk 6]
```diff
@@ -670,3 +589,262 @@ bool _lockFileHasContentHashes(dynamic lockfile) {
...
+      );
+    }
+  }
+  return result;
+}
+
+/// Returns a pubspec with the same dependencies as [original] but with all
+/// version constraints replaced by `>=c` where `c`, is the member of `current`
+/// that has same name as the dependency.
+Pubspec atLeastCurrent(Pubspec original, List<PackageId> current) {
```
[the remark this answers is not in the document]
Comment (sigurdm) [reply lib/src/command/dependency_services.dart:864]: Yeah, not sure why I duplicated it ...

Deleted

**Pull request merged**